File: src/KubeOps.Templates/Templates/Operator.FSharp/Controller/DemoController.fs
namespace GeneratedOperatorProject.Controller

open System
open System.Threading.Tasks
open GeneratedOperatorProject.Entities
open GeneratedOperatorProject.Finalizer
open k8s.Models
open KubeOps.Operator.Controller
open KubeOps.Operator.Controller.Results
open KubeOps.Operator.Finalizer
open KubeOps.Operator.Rbac
open Microsoft.Extensions.Logging

[<EntityRbac(typeof<V1DemoEntity>, Verbs = RbacVerb.All)>]
type DemoController(logger: ILogger<DemoController>, finalizerManager: IFinalizerManager<V1DemoEntity>) =
    interface IResourceController<V1DemoEntity> with
        member this.ReconcileAsync entity =
            async {
                logger.LogInformation $"entity {entity.Name()} called reconcile"

                do! finalizerManager.RegisterFinalizerAsync<DemoFinalizer> entity
                    |> Async.AwaitTask

                return
                    ResourceControllerResult.RequeueEvent
                    <| TimeSpan.FromSeconds 15.0
            }
            |> Async.StartAsTask

        member this.StatusModifiedAsync entity =
            logger.LogInformation $"entity {entity.Name()} called not modified"
            Task.CompletedTask

        member this.DeletedAsync entity =
            logger.LogInformation $"entity {entity.Name()} called deleted"
            Task.CompletedTask

File: _old/src/KubeOps.Templates/Templates/Operator.CSharp/Controller/DemoController.cs
using k8s.Models;
using KubeOps.Operator.Controller;
using KubeOps.Operator.Controller.Results;
using KubeOps.Operator.Finalizer;
using KubeOps.Operator.Rbac;
using GeneratedOperatorProject.Entities;
using GeneratedOperatorProject.Finalizer;

namespace GeneratedOperatorProject.Controller;

[EntityRbac(typeof(V1DemoEntity), Verbs = RbacVerb.All)]
public class DemoController : IResourceController<V1DemoEntity>
{
    private readonly ILogger<DemoController> _logger;
    private readonly IFinalizerManager<V1DemoEntity> _finalizerManager;

    public DemoController(ILogger<DemoController> logger, IFinalizerManager<V1DemoEntity> finalizerManager)
    {
        _logger = logger;
        _finalizerManager = finalizerManager;
    }

    public async Task<ResourceControllerResult?> ReconcileAsync(V1DemoEntity entity)
    {
        _logger.LogInformation($"entity {entity.Name()} called {nameof(ReconcileAsync)}.");
        await _finalizerManager.RegisterFinalizerAsync<DemoFinalizer>(entity);

        return ResourceControllerResult.RequeueEvent(TimeSpan.FromSeconds(15));
    }

    public Task StatusModifiedAsync(V1DemoEntity entity)
    {
        _logger.LogInformation($"entity {entity.Name()} called {nameof(StatusModifiedAsync)}.");

        return Task.CompletedTask;
    }

    public Task DeletedAsync(V1DemoEntity entity)
    {
        _logger.LogInformation($"entity {entity.Name()} called {nameof(DeletedAsync)}.");

        return Task.CompletedTask;
    }
}
